File: app/core/src/main/java/com/fsck/k9/message/extractors/MessageFulltextCreator.java
package com.fsck.k9.message.extractors;


import androidx.annotation.NonNull;

import com.fsck.k9.message.html.HtmlConverter;
import com.fsck.k9.mail.Message;
import com.fsck.k9.mail.Part;
import com.fsck.k9.mail.internet.MessageExtractor;
import com.fsck.k9.mail.internet.MimeUtility;


public class MessageFulltextCreator {
    private static final int MAX_CHARACTERS_CHECKED_FOR_FTS = 200*1024;


    private final TextPartFinder textPartFinder;


    MessageFulltextCreator(TextPartFinder textPartFinder) {
        this.textPartFinder = textPartFinder;
    }

    public static MessageFulltextCreator newInstance() {
        TextPartFinder textPartFinder = new TextPartFinder();
        return new MessageFulltextCreator(textPartFinder);
    }

    public String createFulltext(@NonNull Message message) {
        Part textPart = textPartFinder.findFirstTextPart(message);
        if (textPart == null || hasEmptyBody(textPart)) {
            return null;
        }

        String text = MessageExtractor.getTextFromPart(textPart, MAX_CHARACTERS_CHECKED_FOR_FTS);
        String mimeType = textPart.getMimeType();
        if (!MimeUtility.isSameMimeType(mimeType, "text/html")) {
            return text;
        }

        return HtmlConverter.htmlToText(text);
    }

    private boolean hasEmptyBody(Part textPart) {
        return textPart.getBody() == null;
    }

}

File: legacy/core/src/main/java/com/fsck/k9/message/extractors/MessageFulltextCreator.kt
package com.fsck.k9.message.extractors

import com.fsck.k9.mail.Message
import com.fsck.k9.mail.Part
import com.fsck.k9.mail.internet.MessageExtractor
import com.fsck.k9.mail.internet.MimeUtility
import com.fsck.k9.message.html.HtmlConverter

class MessageFulltextCreator private constructor(private val textPartFinder: TextPartFinder) {
    fun createFulltext(message: Message): String? {
        val textPart = textPartFinder.findFirstTextPart(message)
        if (textPart == null || hasEmptyBody(textPart)) {
            return null
        }

        val text = MessageExtractor.getTextFromPart(textPart, MAX_CHARACTERS_CHECKED_FOR_FTS)
        val mimeType = textPart.mimeType
        return if (!MimeUtility.isSameMimeType(mimeType, "text/html")) {
            text
        } else {
            HtmlConverter.htmlToText(text)
        }
    }

    private fun hasEmptyBody(textPart: Part) = textPart.getBody() == null

    companion object {
        private const val MAX_CHARACTERS_CHECKED_FOR_FTS = 200 * 1024L

        fun newInstance() = MessageFulltextCreator(TextPartFinder())
    }
}
